source/Calamari/Kubernetes/ResourceStatus/ResourceStatusChecker.cs

```diff
@@ -0,0 +1,108 @@
+    
+        // TODO change this to timeout
+        private int count = 20;
+
+        public ResourceStatusChecker(IResourceRetriever resourceRetriever, ILog log)
+        {
+            this.resourceRetriever = resourceRetriever;
+            this.log = log;
+        }
+        
+        public void CheckStatusUntilCompletion(IEnumerable<ResourceIdentifier> resourceIdentifiers, DeploymentContext context, Kubectl kubectl)
+        {
+            var definedResources = resourceIdentifiers.ToList();
+            while (!IsCompleted())
+            {
+                CheckStatus(definedResources, context, kubectl);
+                Thread.Sleep(PollingIntervalSeconds * 1000);
+            }
+        }
+
+        public void CheckStatus(IEnumerable<ResourceIdentifier> definedResources, DeploymentContext context, Kubectl kubectl)
+        {
+            var newResourceStatuses = resourceRetriever
+                .GetAllOwnedResources(definedResources, context, kubectl)
+                .ToDictionary(resource => resource.Uid, resource => resource);
+    
+            var createdOrUpdatedResources = GetCreatedOrUpdatedResources(newResourceStatuses);
+            var removedResources = GetRemovedResources(newResourceStatuses);
+            resources = newResourceStatuses;
+    
+            foreach (var resource in createdOrUpdatedResources.Concat(removedResources))
+            {
+                log.Verbose($"Resource updated: {JsonConvert.SerializeObject(resource)}");
+                SendServiceMessage(resource);
+            }
+        }
+
+        private bool IsCompleted()
+        {
+            return --count < 0 ||
+                   (resources.Count > 0 && resources.All(resource => resource.Value.Status == Resources.ResourceStatus.Successful));
+        }
+
+        private IEnumerable<Resource> GetCreatedOrUpdatedResources(IDictionary<string, Resource> newResourceStatuses)
+        {
+            var createdOrUpdated = new List<Resource>();
+            foreach (var status in newResourceStatuses)
+            {
+                if (!resources.ContainsKey(status.Key) || status.Value.HasUpdate(resources[status.Key]))
+                {
+                    createdOrUpdated.Add(status.Value);
+                }
+            }
+            return createdOrUpdated;
+        }
+
+        private IEnumerable<Resource> GetRemovedResources(IDictionary<string, Resource> newResourceStatuses)
+        {
+            var removed = new List<Resource>();
+            foreach (var resourceEntry in resources)
+            {
+                if (!newResourceStatuses.ContainsKey(resourceEntry.Key))
+                {
+                    resourceEntry.Value.Removed = true;
+                    removed.Add(resourceEntry.Value);
+                }
+            }
+            return removed;
+        }
+
+        private void SendServiceMessage(Resource resource)
+        {
+            // TODO: update this for database
```

Comment:
Do we still need this comment?

Reply:
I actually don't know what it means!

Reply:
The current service message is in the shape that is to be handled as a "DAT" type of log entry. This is in order for the Server to store it in the logs. So it has the Service Message name of "logData", type of "k8s-status" and everything else is in the "data" field.

But once we change to the database solution, I imagine we will implement an IHandleServiceMessage and update database there. In this case, we could use a better schema, for example, the name could be "k8s-resource-status" rather than "logData" which doesn't make too much sense if we don't store it in the logs. Also things like ActionId and Cluster name could be added directly as Service Message properties, so the Server doesn't even need to inspect the content of the data

Reply:
cool, I'm changing it now